src/binary/IterationSpace.cpp

```diff
@@ -0,0 +1,429 @@
+                                                          int64_t i_id_old ){
+
+  int64_t l_direction = (( i_id_old - i_id_new ) + 1) / 2;
+  int64_t l_max_id = i_id_new > i_id_old ? i_id_new : i_id_old;
+  for( int64_t l_di = i_dim_loops.end-1; l_di >= i_dim_loops.begin; l_di-- ){
+    int64_t l_size = m_loop_sizes->at(l_di);
+    if(l_max_id % l_size != 0){
+      return (l_di - m_parallel_loops.begin) * 2 + l_direction;
+    }
+    else{
+      l_max_id /= l_size;
+    }
+  }
+
+  return 0;
+}
+
+int64_t einsum_ir::binary::IterationSpace::getNumTasks( int64_t i_thread_id ){
+  return m_dim_movements[i_thread_id].size();
+}
+
+void einsum_ir::binary::IterationSpace::addMovementOffsets( int64_t          i_thread_id, 
+                                                            int64_t          i_task_id,
+                                                            char    const ** io_ptr_left,
+                                                            char    const ** io_ptr_right,
+                                                            char    const ** io_ptr_out_aux,
+                                                            char          ** io_ptr_out){
+  uint8_t l_move =  m_dim_movements[i_thread_id][i_task_id];
+  int8_t l_direction = 1 - ((l_move & 1) << 1); 
+  l_move = l_move >> 1;
+
+  *io_ptr_left    += l_direction * m_movement_offsets[0][l_move];
+  *io_ptr_right   += l_direction * m_movement_offsets[1][l_move];
+  *io_ptr_out     += l_direction * m_movement_offsets[2][l_move];
+  *io_ptr_out_aux += l_direction * m_movement_offsets[3][l_move];
+}
+
+void einsum_ir::binary::IterationSpace::getInitialOffsets( int64_t   i_thread_id,
+                                                           int64_t & o_off_left,
+                                                           int64_t & o_off_right,
+                                                           int64_t & o_off_out_aux,
+                                                           int64_t & o_off_out ) {
+  o_off_left    = m_initial_offsets[i_thread_id][0];
+  o_off_right   = m_initial_offsets[i_thread_id][1];
+  o_off_out     = m_initial_offsets[i_thread_id][2];
+  o_off_out_aux = m_initial_offsets[i_thread_id][3];
+}
+
+void einsum_ir::binary::IterationSpace::SfcOracle2d( int64_t *i_m, 
+                                                     int64_t *i_n,
+                                                     int64_t *i_omp, 
+                                                     int64_t  i_idx ){
+  
+  int l_w = m_sfc_tasks_m;
+  int l_h = m_sfc_tasks_n;
+  *i_omp = i_idx / (l_w*l_h);
+  i_idx = i_idx % (l_w*l_h);
+
+  int l_idx_m, l_idx_n;
+  gilbert_d2xy(&l_idx_m, &l_idx_n, i_idx, l_w, l_h);
+
+  *i_m = l_idx_m;
+  *i_n = l_idx_n;
+}
+
+#define SIGN(A) (0 < (A) ? (1) : ( 0 == (A) ? (0) : (-1)))
+
+int gilbert_d2xy_r( int dst_idx, int cur_idx,
+                    int *xres, int *yres,
+                    int ax,int ay,
+                    int bx,int by ) {
```

Comment:
Move third-party SFC code to a separate file in a ``third_party`` directory.
Mention the source(s) in a comment, these two?
1. https://github.com/libxsmm/libxsmm-dnn/blob/6ce57235a82d411da428806934e7a6e3aad48625/src/libxsmm_dnn_sfc_tools.c#L64-L67
2. https://github.com/jakubcerveny/gilbert/blob/master/gilbert_d2xy.py

Include licenses of gilbert and LIBXSMM in the separate file.